sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -48,203 +47,231 @@ public class ImplementIDisposableCorrectly : SonarDiagnosticAnalyzer
+                        .FirstOrDefault(IsOrImplementsIDisposable);
+
+                    if (idisposableInterfaceSyntax != null)
+                    {
+                        AddSecondaryLocation(idisposableInterfaceSyntax.GetLocation(),
+                            $"Remove 'IDisposable' from the list of interfaces implemented by '{classSymbol.Name}'"
+                                + " and override the base class 'Dispose' implementation instead.");
+                    }
+
+                    if (FindMethodDeclarationsRecursive(classSymbol.BaseType, IsVirtualDisposeBool).Any())
+                    {
+                        VerifyDisposeOverride(FindMethodDeclarations(classSymbol, IsVirtualDisposeBool)
+                            .OfType<MethodDeclarationSyntax>()
+                            .FirstOrDefault());
+                    }
+
+                    return secondaryLocations;
+                }
+
+                if (classSymbol.Implements(KnownType.System_IDisposable))
+                {
+                    if (!FindMethodDeclarations(classSymbol, IsVirtualDisposeBool).Any())
+                    {
+                        AddSecondaryLocation(classDeclaration.Identifier.GetLocation(),
+                            $"Provide 'protected' overridable implementation of 'Dispose(bool)' on '{classSymbol.Name}'"
+                                + " or mark the type as 'sealed'.");
+                    }
+
+                    var destructor = FindMethodDeclarations(classSymbol, SymbolHelper.IsDestructor)
+                        .OfType<DestructorDeclarationSyntax>()
+                        .FirstOrDefault();
+
+                    VerifyDestructor(destructor);
+
+                    var disposeMethod = FindMethodDeclarations(classSymbol, KnownMethods.IsIDisposableDispose)
+                        .OfType<MethodDeclarationSyntax>()
+                        .FirstOrDefault();
+
+                    VerifyDispose(disposeMethod, hasDestructor: destructor != null);
+                }
+
+                return secondaryLocations;
             }
 
-            var implementationErrors = new List<SecondaryLocation>();
+            private void AddSecondaryLocation(Location location, string message)
+            {
+                secondaryLocations.Add(new SecondaryLocation(location, message));
+            }
 
-            if (classSymbol.BaseType.Implements(KnownType.System_IDisposable))
+            private void VerifyDestructor(DestructorDeclarationSyntax destructorSyntax)
             {
-                var idisposableInterfaceSyntax = classDeclaration.BaseList.Types
-                    .FirstOrDefault(t => IsOrImplementsIDisposable(t, c.SemanticModel));
-                if (idisposableInterfaceSyntax != null)
+                if (destructorSyntax?.Body != null &&
+                    destructorSyntax.Body.ChildNodes().Count() != 1 &&
+                    !CallsVirtualDispose(destructorSyntax.Body, argumentValue: "false"))

```

Comment:
The issue won't be raised if there are != 1 lines and the call to Dispose(bool) is not present.
I think it should be:
destructorSyntax?.Body != null &&
( destructorSyntax.Body.ChildNodes().Count() != 1 || !CallsVirtualDispose(destructorSyntax.Body, argumentValue: "false") )

This is quite difficult to read, perhaps do something like this:
bool hasOneLine = destructorSyntax?.Body != null && destructorSyntax.Body.ChildNodes().Count() == 1;
if (!hasOneLine || !CallsVirtualDispose(...)

Worth also adding a test for where finalizer has more than one line.